lib/service/rpmsg/rpc/rpmsg_rpc_client.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <errno.h>
+#include <metal/mutex.h>
+#include <metal/spinlock.h>
+#include <metal/utilities.h>
+#include <openamp/open_amp.h>
+#include <openamp/rpmsg_rpc_client_server.h>
+#include <stdio.h>
+#include <string.h>
+#include <fcntl.h>
+
+#define LPRINTF(format, ...) printf(format, ##__VA_ARGS__)
+#define LPERROR(format, ...) LPRINTF("ERROR: " format, ##__VA_ARGS__)
+
+void rpmsg_service_client_unbind(struct rpmsg_endpoint *ept)
+{
+	struct rpmsg_rpc_data *rpc;
+
+	rpc = metal_container_of(ept, struct rpmsg_rpc_data, ept);
+	(void)ept;
+	rpmsg_destroy_ept(&rpc->ept);
+	atomic_flag_clear(&rpc->nacked);
+	if (rpc->shutdown_cb)
+		rpc->shutdown_cb(rpc);
+}
+
+int rpmsg_rpc_client_init(struct rpmsg_rpc_data *rpc,
+			  struct rpmsg_device *rdev,
+			  const char *ept_name, uint32_t ept_addr,
+			  uint32_t ept_raddr,
+			  void *poll_arg, rpmsg_rpc_poll poll,
+			  rpmsg_rpc_shutdown_cb shutdown_cb)
+{
+	int ret;
+
+	if (!rpc || !rdev)
+		return -EINVAL;
+	metal_spinlock_init(&rpc->buflock);
+	metal_mutex_init(&rpc->lock);
+	rpc->shutdown_cb = shutdown_cb;
+	rpc->poll_arg = poll_arg;
+	rpc->poll = poll;
+	rpc->reqbuf = NULL;
+	rpc->respbuf = NULL;
+	rpc->respbuf_len = 0;
+	rpc->cb = NULL;
+	atomic_init(&rpc->nacked, 1);
+	ret = rpmsg_create_ept(&rpc->ept, rdev,
+			       ept_name, ept_addr, ept_raddr,
+			       rpmsg_endpoint_client_cb,
+			       rpmsg_service_client_unbind);
+	if (ret != 0) {
+		metal_mutex_release(&rpc->lock);
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+int rpmsg_rpc_client_send(struct rpmsg_rpc_data *rpc,
+			  int rp, void *request_param,
+			  size_t req_param_size, void *response_param,
+			  size_t resp_param_size, void *cb)
```

Comment:
Here the APi inpose that you serialize your RPC request. If you call 2 time this function you will overwrite some rpc structure fileds.
i would suggest you a similar approche on client thatn the serveur using` rpmsg_rpc_services` struct.
at ini you provide the table and when an answer is received, you call the CB depending on the ID.  This would allow to parallelize the requests.